drivers/dma/dma_stm32_v3.c

```diff
@@ -0,0 +1,724 @@
+	dev_config = dev->config;
+
+	if (id >= dev_config->max_channels) {
+		LOG_ERR("Invalid channel ID %d, max channels %d", id, dev_config->max_channels);
+		return -EINVAL;
+	}
+
+	channel_config = &dev_config->channels[id];
+	dma = dev_config->base;
+	channel = dma_stm32_id_to_channel(id);
+
+	if (dma_stm32_disable_channel(dma, channel) != 0) {
+		return -EBUSY;
+	}
+
+	if (channel_config->direction > PERIPHERAL_TO_MEMORY) {
+		LOG_ERR("Invalid DMA direction %d", channel_config->direction);
+		return -EINVAL;
+	}
+
+	LL_DMA_ConfigAddresses(dma, channel, src, dst);
+
+	LL_DMA_SetBlkDataLength(dma, channel, size);
+
+	channel_config->busy = true;
+
+	LL_DMA_EnableChannel(dma, channel);
+
+	return 0;
+}
+
+int dma_stm32_start(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	struct dma_stm32_channel *channel_config;
+	uint32_t channel = dma_stm32_id_to_channel(id);
+	const DMA_TypeDef *dma = dev_config->base;
+
+	if (LL_DMA_IsEnabledChannel(dma, channel)) {
+		LOG_INF("Channel %d is already enabled", id);
+		return 0;
+	}
+
+	/* When starting the dma, the stream is busy before enabling */
+	channel_config = &dev_config->channels[id];
+	channel_config->busy = true;
+
+	dma_stm32_clear_channel_irq(dma, channel);
+
+	LL_DMA_EnableChannel(dma, channel);
+
+	return 0;
+}
+
+int dma_stm32_stop(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	struct dma_stm32_channel *channel_config = &dev_config->channels[id];
+	uint32_t channel = dma_stm32_id_to_channel(id);
+	const DMA_TypeDef *dma = dev_config->base;
+
+	if (channel_config->hal_override) {
+		channel_config->busy = false;
+		return 0;
+	}
+
+	if (!LL_DMA_IsEnabledChannel(dma, channel)) {
+		return 0;
+	}
+
+	dma_stm32_clear_channel_irq(dma, channel);
+
+	dma_stm32_disable_it(dma, channel);
+
+	if (dma_stm32_disable_channel(dma, channel)) {
+		LOG_ERR("Failed to disable DMA channel %d", id);
+		return -EBUSY;
+	}
+
+	channel_config->busy = false;
+
+	return 0;
+}
+
+int dma_stm32_get_status(const struct device *dev, uint32_t id, struct dma_status *status)
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	struct dma_stm32_channel *channel_config;
+	uint32_t channel = dma_stm32_id_to_channel(id);
+	const DMA_TypeDef *dma = dev_config->base;
+
+	if (id >= dev_config->max_channels) {
+		return -EINVAL;
+	}
+
+	channel_config = &dev_config->channels[id];
+
+	status->pending_length = LL_DMA_GetBlkDataLength(dma, channel);
+	status->dir = channel_config->direction;
+	status->busy = channel_config->busy;
+
+	return 0;
+}
+
+int dma_stm32_suspend(const struct device *dev, uint32_t id)
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	uint32_t channel = dma_stm32_id_to_channel(id);
+	const DMA_TypeDef *dma = dev_config->base;
+	uint32_t timeout = 10;
+
+	if (id >= dev_config->max_channels) {
+		return -EINVAL;
+	}
+
+	LL_DMA_SuspendChannel(dma, channel);
+	while (timeout-- > 0) {
```

Comment:
use a WAIT_FOR macro here instead to poll the register status https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#ga68eb35df6b4715714312df90209accee